src/plugins/ds_mongo.c

```diff
@@ -0,0 +1,3400 @@
+ * @param[in] plg_data Plugin data.
+ * @return NULL on success;
+ * @return Sysrepo error info on error.
+ */
+sr_error_info_t *
+srpds_mongo_access_set(const struct lys_module *mod, sr_datastore_t ds, const char *owner, const char *group, mode_t perm, void *plg_data)
+{
+    mongo_data_t mdata;
+    mongo_plg_conn_data_t *pdata = (mongo_plg_conn_data_t *)plg_data;
+
+    sr_error_info_t *err_info = NULL;
+    bson_error_t error;
+    bson_t *bson_query_key = NULL, *bson_query = NULL;
+
+    assert(mod);
+
+    if ((err_info = srpds_data_init(mod, ds, 1, pdata, &mdata))) {
+        goto cleanup;
+    }
+
+    /* _id for owner, group and permissions */
+    bson_query_key = BCON_NEW("_id", "4");
+    
+    /* set owner */
+    if (owner) {
+        bson_query = BCON_NEW("$set", "{", "owner", BCON_UTF8(owner), "}");
+        if (!mongoc_collection_update_one(mdata.module, bson_query_key, bson_query, NULL, NULL, &error)) {
+            ERRINFO(&err_info, SR_ERR_OPERATION_FAILED, "mongoc_collection_update_one()", error.message)
+            goto cleanup;
+        }
+        bson_destroy(bson_query);
+    }
+
+    /* set group */
+    if (group) {
+        bson_query = BCON_NEW("$set", "{", "group", BCON_UTF8(group), "}");
+        if (!mongoc_collection_update_one(mdata.module, bson_query_key, bson_query, NULL, NULL, &error)) {
+            ERRINFO(&err_info, SR_ERR_OPERATION_FAILED, "mongoc_collection_update_one()", error.message)
+            goto cleanup;
+        }
+        bson_destroy(bson_query);
+    }
+
+    /* set permissions */
+    if (perm) {
+        bson_query = BCON_NEW("$set", "{", "perm", BCON_INT32(perm), "}");
+        if (!mongoc_collection_update_one(mdata.module, bson_query_key, bson_query, NULL, NULL, &error)) {
+            ERRINFO(&err_info, SR_ERR_OPERATION_FAILED, "mongoc_collection_update_one()", error.message)
+            goto cleanup;
+        }
+    }
+
+cleanup:
+    bson_destroy(bson_query_key);
+    bson_destroy(bson_query);
+    srpds_data_destroy(pdata, &mdata);
+    return err_info;
+}
+
+/**
+ * @brief Check whether the current user has the required access to datastore data.
+ *
+ * @param[in] mod Specific module.
+ * @param[in] ds Specific datastore.
+ * @param[in] plg_data Plugin data.
+ * @param[out] read Optional, whether the read permission was granted or not.
+ * @param[out] write Optional, whether the write permission was granted or not.
+ * @return NULL on success;
+ * @return Sysrepo error info on error.
+ */
+sr_error_info_t *
+srpds_mongo_access_check(const struct lys_module *mod, sr_datastore_t ds, void *plg_data, int *read, int *write)
+{
+    mongo_data_t mdata;
+    mongo_plg_conn_data_t *pdata = (mongo_plg_conn_data_t *)plg_data;
+
+    sr_error_info_t *err_info = NULL;
+    int is_owner, in_group;
+    char *username = NULL, *groupname = NULL,
+            *owner = NULL, *group = NULL;
+    mode_t perm = 0;
```

Comment:
Variable declarations should always be in a single block.

Reply:
done